### PR TITLE
Tools for our RoboCup docker image

### PR DESCRIPTION
**Pull request by Y0sh1dk**
Currently builds and tags a image locally as `nugus_sim:robocup`. This image contains all runtime depedancies and our compiled binaries. Should be tested with the docker command directly as that is how it will be run for robocup. Currently `nugus_controller` crashes a couple of seconds after connecting for me so cannot test rigerously. This will be used for the image we need to submit on 23rd May. This could probably be split up into multiple PR's later.

Build image:
`./b webots build` (probably a good idea to run with `-c` flag to clean config)

Run container:
`./b webots run webots` (runs `webots` role with `127.0.0.1:10020` as the address of Webots server)
alternatively,
```
docker container run --rm --network=host -e ROBOCUP_ROBOT_ID=1 -e ROBOCUP_TEAM_COLOR=Blue -e ROBOCUP_SIMULATOR_ADDR=127.0.0.1:10020 nugus_sim:robocup webots
```
Key Points:
 - Needs to be run on `host` network
 - Needs to be provided all 3 `ROBOCUP` env vars, or will not run
 - `ROBOCUP_TEAM_COLOR` is not set anywhere as we dont use it?

TODO (before submission):
- Fix `nugus_controller` on Webots side
- Make a role for a simple walk with get-up

TODO (later):
 - target for robocup hardware (`robocup2021` target currently the same as generic, and not  used)
 - log to `/robocup-logs`
 - `./b install` logic can probably be cleaned up and test I didnt break it for installing to a robot
 - Integrate with current dockerise tools?
 - GPU support?

Pls test before submission :sweat_smile:

### CODE REVIEW

File: docker/nugus_sim/run.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    # ROBOCUP_ROBOT_ID
+    file_name = "GlobalConfig.yaml"
+    with open(file_name) as file:
+        data = yaml.load(file)
+    data["player_id"] = int(config["ROBOCUP_ROBOT_ID"])
+
+    with open(file_name, "w") as file:
+        yaml.dump(data, file)
+
+    file_name = "GameController.yaml"
```
Comment (KipHamiltons) [inline docker/nugus_sim/run.py:67]: reusing the same variable names threw me for a loop for a second

File: docker/nugus_sim/run.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    with open(file_name) as file:
+        data = yaml.load(file)
+    data["player_id"] = int(config["ROBOCUP_ROBOT_ID"])
+
+    with open(file_name, "w") as file:
+        yaml.dump(data, file)
+
+    # ROBOCUP_TEAM_COLOR
+    # ??
```
Comment (KipHamiltons) [inline docker/nugus_sim/run.py:76]: If we don't use the team colour, does that mean we have an implicit mapping from team number to colour somewhere? @Bidski 
If so, we should make sure that that mapping is correct for this robocup.

File: .gitignore [hunk 3]
```diff
@@ -36,7 +36,6 @@ packer/build.pwi
 
 # LaTeX stuff
 *.aux
-
```
Comment (KipHamiltons) [inline .gitignore:39]: Unrelated change?

File: docker/usr/local/toolchain/generate_robocup2021_toolchain.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Generate toolchain files for the docker images")
+    parser.add_argument(
+        "--prefix", default=os.path.join("usr", "local"), help="Prefix path to store generated files in"
+    )
+
+    args = parser.parse_args()
+
+    generate_toolchains.generate(args.prefix, "generic", target)
```
Comment (KipHamiltons) [inline docker/usr/local/toolchain/generate_robocup2021_toolchain.py:23]:
```suggestion
    generate_toolchains.generate(args.prefix, "robocup2021", target)
```

File: tools/install.py [hunk 5]
```diff
@@ -36,11 +36,12 @@ def run(target, user, config, toolchain, **kwargs):
...
-        "{}{}".format(k, num): "10.1.1.{}".format(num)
-        for num in range(1, num_robots + 1)
-        for k, v in zip(("nugus", "n", "i", "igus"), [num] * num_robots)
-    }.get(target, target)
+    if target != "local":
+        target = {
+            "{}{}".format(k, num): "10.1.1.{}".format(num)
+            for num in range(1, num_robots + 1)
+            for k, v in zip(("nugus", "n", "i", "igus"), [num] * num_robots)
+        }.get(target, target)
```
Comment (KipHamiltons) [inline tools/install.py:44]: this hurts my brain lol. Not your changes, just this code :joy:

File: tools/webots.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    )
+
+    command_run.add_argument("role", nargs=1, help="Role to run")
+
+
+def run(sub_command, role=None, **kwargs):
+    if sub_command == "build":
+
+        # # Set target
+        # err = subprocess.run(["./b", "target", "robocup2021"]).returncode
```
Comment (KipHamiltons) [inline tools/webots.py:44]: Remove commented code if it doesn't have a purpose

File: tools/webots.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        # if err != 0:
+        #     cprint("returned exit code {}".format(err), "red", attrs=["bold"])
+        #     exit(err)
+
+        # Remove build volume
+        if kwargs["clean"] == True:
+            # Clean config
+            err = subprocess.run(["./b", "configure", "--clean"]).returncode
+            if err != 0:
+                cprint("returned exit code {}".format(err), "red", attrs=["bold"])
```
Comment (KipHamiltons) [inline tools/webots.py:54]:
```suggestion
                cprint("returned exit code {} during NUbots docker volume clean and default configure.".format(err), "red", attrs=["bold"])
```

File: tools/webots.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                exit(err)
+
+        # Set build config
+        err = subprocess.run(
+            ["./b", "configure", "--", "-DCMAKE_BUILD_TYPE=Release", "--", "-DROLE_webots=ON"]
```
Comment (KipHamiltons) [inline tools/webots.py:59]: double `"--"`? I don't follow what the second `"--"` is meant to do

File: tools/webots.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        ).returncode
+        if err != 0:
+            cprint("returned exit code {}".format(err), "red", attrs=["bold"])
```
Comment (KipHamiltons) [inline tools/webots.py:62]:
```suggestion
            cprint("returned exit code {} configuring NUbots.".format(err), "red", attrs=["bold"])
```

File: tools/webots.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+            exit(err)
+
+        # Build code
+        err = subprocess.run(["./b", "build"]).returncode
+        if err != 0:
+            cprint("returned exit code {}".format(err), "red", attrs=["bold"])
```
Comment (KipHamiltons) [inline tools/webots.py:68]:
```suggestion
            cprint("returned exit code {} building NUbots.".format(err), "red", attrs=["bold"])
```

File: tools/webots.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+            exit(err)
+
+        # Copy compiled binaries and runtime dependancies out of build volume
+        err = subprocess.run(["./b", "install", "local", "-t"]).returncode
+        if err != 0:
+            cprint("returned exit code {}".format(err), "red", attrs=["bold"])
```
Comment (KipHamiltons) [inline tools/webots.py:74]:
```suggestion
            cprint("returned exit code {} during copying of binaries to local directory.".format(err), "red", attrs=["bold"])
```

File: tools/webots.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+            exit(err)
+
+        # Build image!
+        os.chdir(b.project_dir + "/docker")
+        err = subprocess.run(
+            ["docker", "build", "-t", "nugus_sim:robocup", "-f", "./nugus_sim_Dockerfile", "."]
+        ).returncode
+        if err != 0:
+            cprint("returned exit code {}".format(err), "red", attrs=["bold"])
```
Comment (KipHamiltons) [inline tools/webots.py:83]:
```suggestion
            cprint("returned exit code {} during build stage of RoboCup2021 docker image.".format(err), "red", attrs=["bold"])
```

**Review by KipHamiltons — COMMENTED** [review]
I think we should make a `config.json` file to test with, so we can test our entrypoint and other stuff
Great progress! It's looking really good to me so far! :grin:

File: .gitignore (hunk 3 above)
Comment (Y0sh1dk) [reply .gitignore]: Unrelated indeed :stuck_out_tongue:

File: tools/install.py (hunk 5 above)
Comment (Y0sh1dk) [reply tools/install.py:44]: Tell me about it :sweat_smile:

File: tools/webots.py (hunk 6 above)
Comment (Y0sh1dk) [reply tools/webots.py:44]: Once a target is made for robocup this will just need to be uncommented out to use it so I think It can stay for now. It can be uncommented now, but It will just build a image the exact same as generic so no point

File: tools/webots.py (hunk 8 above)
Comment (Y0sh1dk) [reply tools/webots.py]: Woops, the second `"--"` shouldnt be there

File: docker/nugus_sim/run.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+import subprocess
+import sys
+from typing import Tuple
+
+import ruamel.yaml
+
+BINARIES_DIR = "/home/nubots/NUbots/binaries"
+CONFIG_DIR = BINARIES_DIR + "/config"
+
+ENV_VARS = {
```
[the remark this answers is not in the document]
Comment (Y0sh1dk) [reply docker/nugus_sim/run.py:13]: The reason I did it this way was to get the unset vars easily [here](https://github.com/NUbots/NUbots/pull/468/files#diff-3c31cc1e636ef5b1f14b4ed9baa005211f5d862419a498154a0642fd6237680dR42), I'm not sure how you would use a list instead to make it cleaner?

File: tools/install.py [hunk 14]
```diff
@@ -49,18 +50,25 @@ def run(target, user, config, toolchain, **kwargs):
...
 
     # Target location to install to
-    target_dir = "{0}@{1}:/home/{0}/".format(user, target)
+    if target != "local":
+        target_binary_dir = "{0}@{1}:/home/{0}/".format(user, target)
+        target_toolchain_dir = "{0}@{1}:/home/{0}/".format(user, target)
+    else:
+        target_binary_dir = b.project_dir + "/docker/nugus_sim/binaries"
+        target_toolchain_dir = b.project_dir + "/docker/nugus_sim/toolchain"
+
```
[the remark this answers is not in the document]
Comment (Y0sh1dk) [reply tools/install.py:59]: It was already `rsync`ing to two different directories (on the robot) before, but because they need to be set to different paths if we are installing locally, it made sense to set them there

File: docker/nugus_sim/run.py (hunk 2 above)
Comment (Bidski) [reply docker/nugus_sim/run.py:78]: This should be set and used somewhere otherwise we are loading the wrong model

File: docker/nugus_sim/run.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (Bidski) [reply docker/nugus_sim/run.py:13]: You dont actually need/use the values in this map as they are just duplicates of the keys

[the remark this answers is not in the document]
Comment (JosephusPaye) [reply docker/nugus_sim/run.py]: Changed to a list
